rfcs/033-api-internal-model-versioning/README.md

```diff
@@ -0,0 +1,78 @@
+
+- If there are problems in the pipeline that delay a reindex, that means a deploy of the API is potentially delayed too.
+
+- If there is a bug in the prod API, fixing it means potentially waiting for a reindex or deploying from a patch.
+
+- We need integration tests that tell us if a build of the API is compatible with the currently available index and we
+  can't guarantee that what is in master is always deployable and that integration tests should always pass.
+
+## Proposed solution
+
+The proposed solution is:
+- for every build in master, release the internal model as an artifact with a new version
+  and publish it to our s3 respository in a similar way as to how we release artifacts in the scala-libs repo.
+
+- all catalogue pipeline services that depend on the internal model keep the dependency on the local module
+  instead of the published artifact. When we work a something that require model changes, we work on our dev branch
+  on model and pipeline services at the same time and once the branch is merged, a new version of the internal model
+  is published.
+
+- the catalogue API depends on the internal model artifact published in s3 instead of the internal model module.
+  We update the version the API depends on in master only when we have an index compatible with that version. Any change
+  in the API that does not require model changes can be merged in master and deployed without pointing the API to a new index.
+
+- If we need to make changes both on the pipeline and on the API, we need to make that in three stages:
+  - change the pipeline
+  - reindex
+  - change the API
+
+## Advantages
+
+This means that:
+- changes to the API that don't need internal model changes can be deployed independently of reindexes
+  and independently of any internal model change.
+
+- If there are problems in the pipeline causing reindex delays, we can still deploy changes to the API,
+  provided that we don't upgrade the version of the internal model dependency.
+
+- We can fix bugs in the API by releasing a new version of the API if we don't need to upgrade model version.
+
+- We can adopt a policy where we say integration tests should always pass and the API should always be deployable
+  from master to production.
+
+- We can split the API into its own repo which could allow us to think substantially differently about the API than
+  the pipeline and have a different build.
+
+
+## Potential problems with the proposed approach
+
+- It's difficult to know from an index what version of the internal model was used to populate it. Ideally we might want to include
+  the model version in the index name but because the pipeline services, including the ingestor, depend on the local internal_model,
+  they don't know the version, so it's not straightforward to do. This means that we could run into issues if we update the index tha API
+  is pointing to without updating the model dependency and viceversa.
```

Comment:
Couple of thoughts:

* The API starts up and discovers the index isn't compatible. Can we make it hard fail, so an ECS blue-green deployment will never allow a broken API to start?
* I wonder if we can copy the version into the catalogue pipeline code? You could put the version in the Scala, something like:

    ```scala
    object InternalModelVersion {
      val version = "1.2.3"
    }
    ```

    and whatever code is responsible for publishing the library to S3 also has to come in and edit this file.

Reply:
> The API starts up and discovers the index isn't compatible. Can we make it hard fail, so an ECS blue-green deployment will never allow a broken API to start?

I think that's probably doable and very sensible

Re: model version & index name I was thinking if we could use the git hash for the internal model version instead of semver. We lose the distinction major minor, patch - which I think we get wrong very often anyway -  but we have a version that is shared between ingestor & internal model and possibly it could be easier to automate a process where the index name has the model version

Reply:
I like that; could we maybe also incorporate the build number from BuildKite?

e.g. `123-cfee1f61b673f1df660ed5eedebca32204e2980c`

Commit hashes are great for knowing where code came from, but lousy for comparisons. It feels like being able to order versions would be useful.

Reply:
ah yes, that sounds like a good idea